price_service/server-rust/src/network/rpc.rs

```diff
@@ -0,0 +1,118 @@
+use {
+    crate::{
+        network::p2p::OBSERVATIONS,
+        Vaa,
+    },
+    anyhow::Result,
+    axum::{
+        routing::get,
+        Router,
+    },
+    dashmap::DashMap,
+    std::sync::Arc,
+};
+
+mod rest;
+
+#[derive(Clone, Default)]
+pub struct VaaCache(Arc<DashMap<String, Vec<(i64, Vaa)>>>);
+
+impl VaaCache {
+    /// Add a VAA to the cache. Keeps the cache sorted by timestamp.
+    fn add(&mut self, key: String, timestamp: i64, vaa: Vaa) -> Result<()> {
+        self.remove_expired()?;
+        let mut entry = self.0.entry(key).or_default();
+        let key = entry
+            .binary_search_by(|(t, _)| t.cmp(&timestamp))
+            .unwrap_or_else(|e| e);
+        entry.insert(key, (timestamp, vaa));
+        Ok(())
+    }
+
+    /// Remove expired VAA's from the cache.
+    fn remove_expired(&mut self) -> Result<()> {
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)?
+            .as_secs() as i64;
+
+        // Scan for items older than now, remove, if the result is empty remove the key altogether.
+        for mut item in self.0.iter_mut() {
+            let (key, vaas) = item.pair_mut();
+            vaas.retain(|(t, _)| t > &now);
+            if vaas.is_empty() {
+                self.0.remove(key);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// For a given set of Price IDs, return the latest VAA for each Price ID.
+    fn latest_for_ids(&self, ids: Vec<String>) -> Vec<(String, Vaa)> {
+        self.0
+            .iter()
+            .filter_map(|item| {
+                if !ids.contains(item.key()) {
+                    return None;
+                }
+
+                let (_, latest_vaa) = item.value().last()?;
+                Some((item.key().clone(), latest_vaa.clone()))
+            })
+            .collect()
```

Comment:
do you want to return an Err here if you can't find one of the ids? It might be surprising for callers of `latest_vaas` to get back an update that contains only a subset of the data they requested. I'm not sure if that logic makes sense here or in the endpoint itself.